### PR TITLE
RT-1.32: BGP policy actions - MED, LocPref, prepend, flow-control

### PR DESCRIPTION
**Pull request by rszarecki**
Test to validate BGP actions: MED

### CODE REVIEW

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 1]
```diff
@@ -0,0 +1,599 @@
...
+*   Configure an route-policy definition with the name ```med-policy```
+    *   /routing-policy/policy-definitions/policy-definition/config/name
+*   For routing-policy ```med-policy``` configure a statement with the name ```match-statement-1```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/config/name
+*   For routing-policy ```med-policy``` statement ```match-statement-1``` set policy-result as ```ACCEPT_ROUTE```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/config/policy-result
+*   For routing-policy ```med-policy``` statement ```match-statement-1``` set MED as ```100```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/config/set-med
+##### Configure  bgp import and export policy for the DUT IPv4 and IPv6 BGP neighbors on ATE Port-1
+*   Set default import and export policy to ```REJECT_ROUTE```
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:53]:
```suggestion
*   Set default import and export policy to `REJECT_ROUTE`.   (Note: even though this is the OC default, the DUT should still accept this configuration)

```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 2]
```diff
@@ -0,0 +1,599 @@
...
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-import-policy
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-export-policy
+*   Apply policy as import and export as a chain/list ```[med-policy]```
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:56]:
```suggestion
*   Add `policy-definition["med-policy"]` to import-policy and export-policy leaf-lists.
```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 3]
```diff
@@ -0,0 +1,599 @@
...
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-import-policy
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-export-policy
+##### Verification
+*   Verify that policies are successfully applied to the DUT BGP neighbor on ATE Port-1 and default policies are set to ```REJECT_ROUTE```
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/state/import-policy
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/state/export-policy
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-import-policy
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-export-policy
+*   Verify that there is no policies applied to the DUT BGP neighbor on ATE Port-2 and default policies are set to ```ACCEPT_ROUTE```
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:69]:
```suggestion
*   Verify that there are no policies applied to the DUT BGP neighbor on ATE Port-2 and default policies are set to ```ACCEPT_ROUTE```
```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 4]
```diff
@@ -0,0 +1,599 @@
...
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/config/name
+*   For routing-policy ```med-policy``` statement ```match-statement-1``` set policy-result as ```ACCEPT_ROUTE```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/config/policy-result
+*   For routing-policy ```med-policy``` statement ```match-statement-1``` set MED as ```100```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/config/set-med
+##### Configure  bgp import and export policy for the DUT IPv4 and IPv6 BGP neighbors on ATE Port-2
+*   Set default import and export policy to ```REJECT_ROUTE```
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-import-policy
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-export-policy
+*   Apply policy as import and export as a chain/list ```[med-policy]```
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:96]:
```suggestion
*   Add `policy-definition["med-policy"]` to import-policy and export-policy leaf-lists.
```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 5]
```diff
@@ -0,0 +1,599 @@
...
+    *   /network-instances/network-instance/protocols/protocol/bgp/global/afi-safis/afi-safi/config/
+    *   Advertise ```ipv4-network-1 = 192.168.10.0/24``` and ```ipv6-network-1 = 2024:db8:128:128::/64``` from ATE to DUT over the IPv4 and IPv6 eBGP session on port-1 with:
+        * MED = 50
+        * Local Preference = 50
+*   Configure IPv4 and IPv6 eBGP between DUT Port-2 and ATE Port-2
+    *   /network-instances/network-instance/protocols/protocol/bgp/global/config
+    *   /network-instances/network-instance/protocols/protocol/bgp/global/afi-safis/afi-safi/config/
+    *   Advertise ```ipv4-network-2 = 192.168.20.0/24``` and ```ipv6-network-2 = 2024:db8:64:64::/64``` from ATE to DUT over the IPv4 and IPv6 eBGP session on port-2 with:
+        * MED = 50
+        * Local Preference = 50
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:38]: I suggest clarifying that the ATE is setting these BGP attributes (vs. the DUT which is what is done in the tests below).

```suggestion
    *   Advertise ```ipv4-network-2 = 192.168.20.0/24``` and ```ipv6-network-2 = 2024:db8:64:64::/64``` from ATE to DUT over the IPv4 and IPv6 eBGP session on port-2.  The ATE should advertise both prefixes with:
        * MED = 50
        * Local Preference = 50
```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 6]
```diff
@@ -0,0 +1,599 @@
...
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/config/name
+*   For routing-policy ```med-policy``` statement ```match-statement-1``` set policy-result as ```ACCEPT_ROUTE```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/config/policy-result
+*   For routing-policy ```med-policy``` statement ```match-statement-1``` set MED as ```+100```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/config/set-med
+##### Configure  bgp import and export policy for the DUT IPv4 and IPv6 BGP neighbors on ATE Port-1
+*   Set default import and export policy to ```REJECT_ROUTE```
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-import-policy
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-export-policy
+*   Apply policy as import and export as a chain/list ```[med-policy]```
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:139]:
```suggestion
*   Add `policy-definition["med-policy"]` to import-policy and export-policy leaf-lists.
* ```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 7]
```diff
@@ -0,0 +1,599 @@
...
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/config/name
+*   For routing-policy ```med-policy``` statement ```match-statement-1``` set policy-result as ```ACCEPT_ROUTE```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/config/policy-result
+*   For routing-policy ```med-policy``` statement ```match-statement-1``` set MED as ```+100```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/config/set-med
+##### Configure  bgp import and export policy for the DUT IPv4 and IPv6 BGP neighbors on ATE Port-2
+*   Set default import and export policy to ```REJECT_ROUTE```
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-import-policy
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-export-policy
+*   Apply policy as import and export as a chain/list ```[med-policy]```
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:182]:
```suggestion
*   Add `policy-definition["med-policy"]` to import-policy and export-policy leaf-lists.
```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 8]
```diff
@@ -0,0 +1,599 @@
...
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/config/name
+*   For routing-policy ```lp-policy``` statement ```match-statement-1``` set policy-result as ```ACCEPT_ROUTE```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/config/policy-result
+*   For routing-policy ```lp-policy``` statement ```match-statement-1``` set Local Preference as ```100```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/config/set-local-pref
+##### Configure  bgp import and export policy for the DUT IPv4 and IPv6 BGP neighbors on ATE Port-1
+*   Set default import and export policy to ```REJECT_ROUTE```
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-import-policy
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-export-policy
+*   Apply as import and export only policy - ```[lp-policy]```
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:225]:
```suggestion
*   Add `policy-definition["lp-policy"]` to import-policy and export-policy leaf-lists.
```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 9]
```diff
@@ -0,0 +1,599 @@
...
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/config/name
+*   For routing-policy ```lp-policy``` statement ```match-statement-1``` set policy-result as ```ACCEPT_ROUTE```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/config/policy-result
+*   For routing-policy ```lp-policy``` statement ```match-statement-1``` set Local Preference as ```100```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/config/set-local-pref
+##### Configure  bgp import and export policy for the DUT IPv4 and IPv6 BGP neighbors on ATE Port-2
+*   Set default import and export policy to ```REJECT_ROUTE```
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-import-policy
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-export-policy
+*   Apply as import and export only policy - ```[lp-policy]```
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:269]:
```suggestion
*    Add `policy-definition["lp-policy"]` to import-policy and export-policy leaf-lists.
```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 10]
```diff
@@ -0,0 +1,599 @@
...
+---
+##### Configure a route-policy to prepend 10
+*   Configure an route-policy definition with the name ```prepend-policy```
+    *   /routing-policy/policy-definitions/policy-definition/config/name
+*   For routing-policy ```prepend-policy``` configure a statement with the name ```match-statement-1```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/config/name
+*   For routing-policy ```prepend-policy``` statement ```match-statement-1``` set policy-result as ```ACCEPT_ROUTE```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/config/policy-result
+*   For routing-policy ```prepend-policy``` statement ```match-statement-1``` prepend as-path with local ASN  ```10``` times
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/set-as-path-prepend/config/repead-n
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:307]:
```suggestion
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/set-as-path-prepend/config/repeat-n
```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 11]
```diff
@@ -0,0 +1,599 @@
...
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:350]:
```suggestion
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/set-as-path-prepend/config/repeat-n
```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 12]
```diff
@@ -0,0 +1,599 @@
...
+---
+##### Configure a route-policy to prepend 10
+*   Configure an route-policy definition with the name ```prepend-policy```
+    *   /routing-policy/policy-definitions/policy-definition/config/name
+*   For routing-policy ```prepend-policy``` configure a statement with the name ```match-statement-1```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/config/name
+*   For routing-policy ```prepend-policy``` statement ```match-statement-1``` set policy-result as ```ACCEPT_ROUTE```
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/config/policy-result
+*   For routing-policy ```prepend-policy``` statement ```match-statement-1``` prepend as-path with ```23456``` ASN  ```10``` times
+    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/set-as-path-prepend/config/repead-n
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:393]:
```suggestion
    *   /routing-policy/policy-definitions/policy-definition/statements/statement/actions/bgp-actions/set-as-path-prepend/config/repeat-n
```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 13]
```diff
@@ -0,0 +1,599 @@
...
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-import-policy
+    *   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/config/default-export-policy
+#### Validate test results
+*   Validate that the ATE receives the prefix ```ipv4-network-1```  from DUT neighbor on ATE Port-2 and it has 11 ASN on as-path. First 10 equial to ```23456``` and last equal to DUT's ASN.
+*   Validate that the ATE receives the prefix ```ipv6-network-1```  from DUT neighbor on ATE Port-2 and it has 11 ASN on as-path. First 10 equial to ```23456``` and last equal to DUT's ASN.
+*   Validate that the ATE receives the prefix ```ipv4-network-2```  from DUT neighbor on ATE Port-1 and it has 11 ASN on as-path. First equial to ATE port-2 ASN and other 10 equal to ```23456``` ASN.
+*   Validate that the ATE receives the prefix ```ipv6-network-2```  from DUT neighbor on ATE Port-1 and it has 11 ASN on as-path. First equial to ATE port-2 ASN and other 10 equal to ```23456``` ASN.
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:424]:
```suggestion
*   Validate that the ATE receives the prefix ```ipv4-network-1```  from DUT neighbor on ATE Port-2 and it has 11 ASN on as-path. First 10 equal to ```23456``` and last equal to DUT's ASN.
*   Validate that the ATE receives the prefix ```ipv6-network-1```  from DUT neighbor on ATE Port-2 and it has 11 ASN on as-path. First 10 equal to ```23456``` and last equal to DUT's ASN.
*   Validate that the ATE receives the prefix ```ipv4-network-2```  from DUT neighbor on ATE Port-1 and it has 11 ASN on as-path. First equal to ATE port-2 ASN and other 10 equal to ```23456``` ASN.
*   Validate that the ATE receives the prefix ```ipv6-network-2```  from DUT neighbor on ATE Port-1 and it has 11 ASN on as-path. First equal to ATE port-2 ASN and other 10 equal to ```23456``` ASN.
```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 14]
```diff
@@ -0,0 +1,599 @@
...
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:468]:
```suggestion
*   Validate that the ATE receives the prefix ```ipv4-network-1```  from DUT neighbor on ATE Port-2 and it has 11 ASN on as-path. First 10 equal to ```23456``` and last equal to DUT's ASN.
*   Validate that the ATE receives the prefix ```ipv6-network-1```  from DUT neighbor on ATE Port-2 and it has 11 ASN on as-path. First 10 equal to ```23456``` and last equal to DUT's ASN.
*   Validate that the ATE receives the prefix ```ipv4-network-2```  from DUT neighbor on ATE Port-1 and it has 11 ASN on as-path. First equal to ATE port-2 ASN and other 10 equal to ```23456``` ASN.
*   Validate that the ATE receives the prefix ```ipv6-network-2```  from DUT neighbor on ATE Port-1 and it has 11 ASN on as-path. First equal to ATE port-2 ASN and other 10 equal to ```23456``` ASN.
```

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 15]
```diff
@@ -0,0 +1,599 @@
...
+*   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/state/import-policy
+*   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/state/export-policy
+*   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/state/default-import-policy
+*   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/apply-policy/state/default-export-policy
+
+## Protocol/RPC Parameter Coverage
+
+* gNMI
+  * Get
+  * Set
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:595]:
```suggestion
  * Subscribe (ONCE)
  * Set (REPLACE)
```

**Review by dplore — CHANGES_REQUESTED** [review]

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 16]
```diff
@@ -0,0 +1,599 @@
...
+- Verify abilty to prepend AS path with 10 additional repetitions of configured ASN in export and import policy
+- verify ```NEXT-STATEMENT``` flow-control action
+- Applicable to both IPv4 and IPv6 BGP neighbors
+
+## Testbed type
+
+*   https://github.com/openconfig/featureprofiles/blob/main/topologies/atedut_2.testbed
+
+## Procedure
+
```
[the remark this answers is not in the document]
Comment (rszarecki) [reply feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:18]: I do not know if this is safe, as it impacts interface configuration. Hence can effectivly remove management access.

Instead I add above instruction under each tets, when configuration modifies only policy tree.

File: feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md [hunk 17]
```diff
@@ -0,0 +1,605 @@
...
+## Testbed type
+
+*   https://github.com/openconfig/featureprofiles/blob/main/topologies/atedut_2.testbed
+
+## Procedure
+### Applying configuration
+
+For each section of configuration below, prepare a gnmi.SetBatch  with all the configuration items appended to one SetBatch.  Then apply the configuration to the DUT in one gnmi.Set using the `replace` option.
+> WARNING: Ensure that replacement do not remove pre-existing configuration objects necessary for DUT manage,ent access and base operation.
```
Comment (dplore) [inline feature/bgp/policybase/otg_tests/actions-MED_LocPref_prepend_flow-control/README.md:21]:
```suggestion
> WARNING: Replace operations should be performed at an appropriate level in the config tree to ensure that preexisting configuration objects necessary for DUT management access and base operation are not removed.
```

**Review by dplore — APPROVED** [review]
One last suggestion regarding the warning on gnmi replace

**Review by dplore — APPROVED** [review]

**Pull request merged**